src/python/src/grpc/_cython/_cygrpc/completion_queue.pyx

```diff
@@ -0,0 +1,152 @@
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+cimport cpython
+
+from grpc._cython._cygrpc cimport call
+from grpc._cython._cygrpc cimport records
+
+import threading
+import time
+
+
+cdef class _CompletionQueuePollFuture:
+  # Supports (most) of the Python 3 Future surface. Timeouts are not supported
+  # as they're specified by the completion queue.
+
+  cdef CompletionQueue completion_queue
+  cdef records.Event event_result
+  cdef object thread
+
+  @staticmethod
+  def _poll(CompletionQueue completion_queue not None,
+            records.Timespec deadline=None):
+    future = _CompletionQueuePollFuture()
+    def poller_function():
+      future.set_result(completion_queue.poll(deadline))
+    future.thread = threading.Thread(target=poller_function)
+    future.thread.start()
+    return future
+
+  cdef set_result(self, records.Event result):
+    self.event_result = result
+
+  def result(self):
+    self.thread.join()
+    return self.event_result
+
+cdef class CompletionQueue:
+
+  def __cinit__(self):
+    self.c_completion_queue = grpc.grpc_completion_queue_create()
+    self.is_shutting_down = False
+    self.is_shutdown = False
+    self.poll_condition = threading.Condition()
+    self.is_polling = False
+
+  def poll(self, records.Timespec deadline=None):
+    # We name this 'poll' to avoid problems with CPython's expectations for
+    # 'special' methods (like next and __next__).
+    cdef grpc.gpr_timespec c_deadline = grpc.gpr_inf_future
+    cdef records.OperationTag tag = None
+    cdef object user_tag = None
+    cdef call.Call operation_call = None
+    cdef records.CallDetails request_call_details = None
+    cdef records.Metadata request_metadata = None
+    cdef records.Operations batch_operations = None
+    if deadline is not None:
+      c_deadline = deadline.c_time
+    cdef grpc.grpc_event event
+
+    # Poll within a critical section
+    with self.poll_condition:
+      while self.is_polling:
+        self.poll_condition.wait(float(deadline) - time.time())
+      self.is_polling = True
+    with nogil:
+      event = grpc.grpc_completion_queue_next(
+          self.c_completion_queue, c_deadline)
+    with self.poll_condition:
+      self.is_polling = False
+      self.poll_condition.notify()
+
+    if event.type == grpc.GRPC_QUEUE_TIMEOUT:
+      return records.Event(event.type, False, None, None, None, None, None)
+    elif event.type == grpc.GRPC_QUEUE_SHUTDOWN:
+      self.is_shutdown = True
+      return records.Event(event.type, True, None, None, None, None, None)
+    else:
+      if event.tag != NULL:
+        # TODO(atash): Does this work? Do I need to <object> *then*
```

Comment:
It's leftover cruft from before I'd written the tests. Angle brackets are used to denote casting operators in Cython the same way parentheses are used in C.